src/server/injected/injectedScript.ts

```diff
@@ -156,24 +156,38 @@ export class InjectedScript {
   }
 
   private _createTextEngine(shadow: boolean): SelectorEngine {
+    const queryList = (root: SelectorRoot, selector: string, single: boolean): Element[] => {
+      const { matcher, strict } = createTextMatcher(selector);
+      const result: Element[] = [];
+      let lastDidNotMatchSelf: Element | null;
+
+      const checkElement = (element: Element) => {
+        if (!strict && lastDidNotMatchSelf && lastDidNotMatchSelf.contains(element))
```

Comment:
FYi: contains doesn't work on shadow roots, so this is an opportunity for further improvement